g3/lowpan6_g3.c

```diff
@@ -121,9 +121,213 @@ static struct lowpan6_g3_data lowpan6_data = {
+  for (i = 0; i < LOWPAN6_ROUTING_TABLE_SIZE; i++) {
+    if (lowpan6_data.routing_table[i].valid_time == 0) {
+      lowpan6_data.routing_table[i].dest_addr = dst;
+      lowpan6_data.routing_table[i].next_addr = next;
+      lowpan6_data.routing_table[i].metric = metric;
+      lowpan6_data.routing_table[i].hop_count = hop_count;
+      lowpan6_data.routing_table[i].metric_type = lowpan6_data.metric_type;
+      lowpan6_data.routing_table[i].valid_time = lowpan6_data.routing_table_ttl;
+      lowpan6_data.routing_table[i].seq_num = 0xFFFF;
+      lowpan6_data.routing_table[i].is_bidirectional = 1;
+      lowpan6_data.n_routing_entries++;
+      return &lowpan6_data.routing_table[i];
+    }
+  }
 
-/* Fragmentation specific functions: */
+  return NULL;
+}
+
+/**
+ * Function used for routing table management.
+ */
+struct lowpan6_g3_routing_entry *
+lowpan6_g3_routing_table_lookup(u16_t dst, u8_t bidirectional_only)
+{
+  unsigned i;
+
+  for (i = 0; i < LOWPAN6_ROUTING_TABLE_SIZE; i++) {
+    if (lowpan6_data.routing_table[i].valid_time > 0 &&
+        lowpan6_data.routing_table[i].dest_addr == dst &&
+        (!bidirectional_only || lowpan6_data.routing_table[i].is_bidirectional))
+      return &lowpan6_data.routing_table[i];
+  }
+
+  return NULL;
+}
+
+/**
+ * Function used for table lookup when routing packets.
+ * Returns ERR_OK if a route is found and sets 'next' pointer
+ * to the next hop address.
+ */
+err_t
+lowpan6_g3_routing_table_route(struct lowpan6_link_addr *dst, struct lowpan6_link_addr *next)
+{
+  struct lowpan6_g3_routing_entry *entry;
+
+  if ((entry = lowpan6_g3_routing_table_lookup(lowpan6_link_addr_to_u16(dst), 1)) == NULL) {
+    return ERR_VAL;
+  }
+
+  lowpan6_link_addr_set_u16(next, entry->next_addr);
+
+  return ERR_OK;
+}
+
+void
+lowpan6_g3_routing_table_delete(struct lowpan6_g3_routing_entry *entry)
+{
+  if (entry == NULL)
+    return;
+  entry->valid_time = 0;
+  lowpan6_data.n_routing_entries--;
+}
+
+static int
+lowpan6_g3_broadcast_log_table_add(u16_t src_addr, u8_t seq_num)
+{
+  unsigned i;
+
+  for (i = 0; i < LOWPAN6_BROADCAST_LOG_TABLE_SIZE; i++) {
+    if (lowpan6_data.broadcast_log_table[i].valid_time == 0) {
+      lowpan6_data.broadcast_log_table[i].src_addr = src_addr;
+      lowpan6_data.broadcast_log_table[i].seq_number = seq_num;
+      lowpan6_data.broadcast_log_table[i].valid_time = lowpan6_data.broadcast_log_table_ttl;
+
+      return i;
+    }
+  }
+
+  return -1;
+}
+
+static struct lowpan6_g3_broadcast_log_entry *
+lowpan6_g3_broadcast_log_table_lookup(u16_t src_addr, u8_t seq_num)
+{
+  unsigned i;
+
+  for (i = 0; i < LOWPAN6_BROADCAST_LOG_TABLE_SIZE; i++) {
+    if (lowpan6_data.broadcast_log_table[i].valid_time > 0 &&
+        lowpan6_data.broadcast_log_table[i].src_addr == src_addr &&
+        lowpan6_data.broadcast_log_table[i].seq_number == seq_num)
+          return &lowpan6_data.broadcast_log_table[i];
+  }
+
+  return NULL;
+}
+
+int
+lowpan6_g3_group_table_add(u16_t addr)
```

Comment:
This *could* be `static`, but now I use to for testing. Once we find out, how to manage the G3-ADP on Phoenix-RTOS it could be made `static`.